### ISSUE TITLE
DI for engine and console runner

### ISSUE THREAD

**Issue by NikolayPianikov** [nunit/nunit-console#1335]
Hey @CharliePoole. What if NUnit projects like engine, console etc. will use some DI. Have you considered this possibility or is it superfluous there? I can suggest a DI based on source code generators that is compatible with NET 2.0+, doesn't add dependencies, is simple and works very fast.

**Comment by NikolayPianikov** [comment]
I think development and testing of NUnit itself will become easier

**Comment by NikolayPianikov** [comment]
Yes, I know, but you have a lot of experience with all parts of NUnit. I meant [this] (https://github.com/DevTeam/Pure.DI) project. Briefly why is suitable for NUnit:
- Doesn't add any dependencies and all generated code is internal (we can control this option)
- Supports all platforms because it does not use reflection or other tricks - just a set of constructors
- Supports all frameworks since Full .NET 2.0, but we can also support the first version - .NET 1.0 if necessary.
- Builds code in a pure DI approach, if we wrote it manually, so we can refuse to use it at any time, since in fact it is pure DI
- Guarantees the correctness of the object graph already at the compilation stage, so errors at runtime are impossible
- Runs very fast - 0 memory and performance overhead, the actual generated code is a chain of constructors

**Comment by OsirisTerje** [comment]
Where was this discussed before?   Issue ?

**Comment by OsirisTerje** [comment]
Thanks @stevenaw!   I see from #346 that @rprouse started this back in 2014, and there are a lot of thoughts around it.  So would be great if @rprouse could chime in here.  Just said, I am all in favor of having DI here and in the framework (or wherever really), but we need to be sure we do this so that it will last.
Rob also raised the thought that it could leverage other types of DI too.  For external users that might be an idea, but for internal use we would need a chosen implementation. 
Since this is a source generator approach it could work, it would mean there would be no external dependencies if I understand it correctly.   But what will happen over the years?

**Comment by NikolayPianikov** [comment]
> Since this is a source generator approach it could work, it would mean there would be no external dependencies if I understand it correctly.

Yes, you are right.

> But what will happen over the years?

This is a reasonable question.
- The version of the source code generator is defined in the NUnit project, i.e. it will work regardless of any changes for the next version of the source code generator.
- NUnit support will be given high priority.
- The proposed project is open and you can always make a fork.
I can share this GitHub project and NuGet repo for publishing packages with members of the NUnit team.
- Since it's pure DI, you can just copy the generated code and maintain it manually. It is important to note that the generated code is as simple as possible.

**Comment by CharliePoole** [comment]
Maybe the next step is to imagine a number of use cases for DI in the engine, framework, adapter.

**Comment by NikolayPianikov** [comment]
Yes, we need to decide on the scenarios, but for sure we will not take into account all of them. In this regard, I would like to note that all scenarios supported by competitors are supported. It must be remembered that this source generator creates an object graph at compile time, i.e. you can’t add something to the graph at runtime - this is a task, not a DI task. For example, you cannot iterate through some assembly, search for all interface implementations and add bindings dynamically - this is not a DI task, as I already noted. The object graph must be defined in advance. If you need to be able to create many objects (replicator) then the graph remains predefined, you just need to inject a `Func<T>`

**Comment by NikolayPianikov** [comment]
Once all the people involved have (or don't) come to a consensus on this issue, I, for my part, can take some recommended part and use DI there to see if it fits and satisfies the requirements. For the purity of the experiment, it would be better if this part was not very large, but required the use of all the features that would be needed in other parts of the project. Then we can better understand how successful the experiment was, come to the conclusion whether it is worth continuing to move in this direction. What do you think about it?

**Comment by OsirisTerje** [comment]
@SeanKilleen  You have comments?

**Comment by SeanKilleen** [comment]
Hi @NikolayPianikov! Sorry for the delay after @OsirisTerje tagged me in.

While I think DI as described could work, in order for us to be able to take such a dependency, we'd likely need to think very hard about which DI library to use, given how we fit into software supply chains. Even though DI may be a good idea, we'd need to consider how widespread the usage is, the level of active support and maintenance for a library, and its existing longevity. So while I love the suggestion, and appreciate the work you put into this particular DI solution, it's hard for me to suggest for us to consider this library vs the existing contenders for such a solution. I think it requires more discussion and a deeper think on our end.

**Comment by OsirisTerje** [comment]
I'll close this for now, and we can come back to this sometime in the future :-)

**Comment by NikolayPianikov** [comment]
@SeanKilleen As I mentioned above, I offer NOT a library, but a source code generator that builds code in a pure DI approach, if we wrote it manually. In fact it is pure DI, without any tricks:

```c#
new Abc (new Xyz(), ...)
```

This allows you to refuse to use it at any time, and do everything manually, as it currently works. I understand that it is very difficult to break stereotypes. In fact, I suggested writing code in a pure DI paradigm and using an assistant to do routine work that is already being done. But anyway, thanks for your time :)
